env/env_test.cc

```diff
@@ -579,142 +593,127 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
   ASSERT_EQ(1U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
-  ASSERT_TRUE(tasks[0].IsSleeping());
-  ASSERT_TRUE(!tasks[1].IsSleeping());
-  ASSERT_TRUE(!tasks[2].IsSleeping());
+  ASSERT_TRUE(tasks[0].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[1].ShortWaitForWokenUp());
+  ASSERT_TRUE(tasks[2].ShortWaitForWokenUp());
 
   // Increase to 5 threads. Task 0 and 2 running.
   env_->SetBackgroundThreads(5, Env::Priority::HIGH);
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ((unsigned int)0, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
-  ASSERT_TRUE(tasks[0].IsSleeping());
-  ASSERT_TRUE(tasks[2].IsSleeping());
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
+  ASSERT_TRUE(tasks[0].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[2].ShortWaitForSleeping());
 
   // Change number of threads a couple of times while there is no sufficient
   // tasks.
   env_->SetBackgroundThreads(7, Env::Priority::HIGH);
   Env::Default()->SleepForMicroseconds(kDelayMicros);
   tasks[2].WakeUp();
-  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
   env_->SetBackgroundThreads(3, Env::Priority::HIGH);
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
   env_->SetBackgroundThreads(4, Env::Priority::HIGH);
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
   env_->SetBackgroundThreads(5, Env::Priority::HIGH);
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
   env_->SetBackgroundThreads(4, Env::Priority::HIGH);
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
-
-  Env::Default()->SleepForMicroseconds(kDelayMicros * 50);
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
 
   // Enqueue 5 more tasks. Thread pool size now is 4.
   // Task 0, 3, 4, 5 running;6, 7 waiting.
   for (size_t i = 3; i < 8; i++) {
     env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &tasks[i],
                    Env::Priority::HIGH);
   }
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ(2U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
-  ASSERT_TRUE(tasks[3].IsSleeping());
-  ASSERT_TRUE(tasks[4].IsSleeping());
-  ASSERT_TRUE(tasks[5].IsSleeping());
-  ASSERT_TRUE(!tasks[6].IsSleeping());
-  ASSERT_TRUE(!tasks[7].IsSleeping());
+  SleepUntilQueueLen(Env::Priority::HIGH, 2U);
+  ASSERT_TRUE(tasks[3].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[4].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[5].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[6].ShortWaitForWokenUp());
+  ASSERT_TRUE(tasks[7].ShortWaitForWokenUp());
```

Comment:
task 6 and 7 do not need to call wait. We can directly assert their states are `!sleep_`.